File: SortingAlgorithm/quickSort.cpp
//picking the last element as pivot
//O(nlogn) - average case running time
//O(n^2) - worst case but it can be avoided using ramdomized version
//randomized version gives high probability of o(nlogn)
using namespace std;
class Solution
{
public:
    //Function to sort an array using quick sort algorithm.
    void quickSort(int arr[], int low, int high)
    {
        // code here
        if (low < high){
            int pindex= partition(arr, low, high);
            quickSort(arr, low, pindex -1);
            quickSort(arr, pindex + 1, high);
        }
    }

public:
    int partition (int arr[], int low, int high)
    {
        // Your code here
        int pindex = 0;
        int pivot = arr[high];
        for (int i = 0; i < high; i++){
            if (arr[i] <= pivot){
                int temp = arr[i];
                arr[i] = arr[pindex];
                arr[pindex] = temp;
                pindex++;
            }
        }
        int temp = arr[pindex];
        arr[pindex] = arr[high];
        arr[high] = temp;
        return pindex;
    }
};
## Randomized quicksort
// suppose we  all have  repeating character in array [4, 4, 4, 4, 4, 4]. if we use single partition method, in the first partition we will get 2 groups, one is empty and
//another is n-1, the next will be n-2, n-3  and so on which makes O(n^2). We can can handle duplicate keys by using three ways partitioning method such as Dutch national flag
class Solution
{
public:
    //Function to sort an array using quick sort algorithm.
    void quickSort(int arr[], int low, int high)
    {
        // code here
        if (low < high){
            int pindex= RandomizedPartition(arr, low, high);
            quickSort(arr, low, pindex -1);
            quickSort(arr, pindex + 1, high);
        }
    }

public:


    int RandomizedPartition(int arr[], int low, int high){
        srand(time(nullptr));
        int random = low + rand() % (high - low);
        int pindex = 0;
        int pivot = arr[high];
        for (int i = 0; i < high; i++){
            if (arr[i] <= pivot){
                int temp = arr[i];
                arr[i] = arr[pindex];
                arr[pindex] = temp;
                pindex++;
            }
        }
        int temp = arr[pindex];
        arr[pindex] = arr[high];
        arr[high] = temp;
        return pindex;
    }
};
//improved quicksort to handle repeating keys
class Solution
{
public:
    //Function to sort an array using quick sort algorithm.
    void quickSort(int arr[], int low, int high)
    {
        // code here
        if (low < high){
            pair<int, int> p = dNL_partition(arr, low, high);
            quickSort(arr, low, p.first-1);
            quickSort(arr, p.second + 1, high);
        }
    }

public:

    pair<int, int> dNL_partition(int arr[], int low, int high){
        srand(time(nullptr));
        int random = low + rand() % (high - low);
        int pivot = arr[high];
        int firstPIndex = low;
        //forward pass to check if arr[i] is less than the pivot
        for(int i = low; i < high; i++){
            if(arr[i] < pivot){
                int temp = arr[i];
                arr[i] = arr[firstPIndex];
                arr[firstPIndex] = temp;
                firstPIndex += 1;
            }
        }
        int secondPINdex = high;
        //backward pass to check if arr[i] is greater than pivot
        // if we get arr[i] < pivot, it means we are in the lesser region. In this case, we break the loop
        for(int i = high; i >= low; i--){
            if (arr[i] > pivot){
                int temp = arr[i];
                arr[i] = arr[secondPINdex];
                arr[secondPINdex] = temp;
                secondPINdex--;
            }
            else if (arr[i] < pivot){
                break;
            }
        }
        return {firstPIndex, secondPINdex};
    }
};

File: SortingAlgorithm/quick sort.py
# quick sort algorithm
from random import randint


def swap(arr, left, right):
    temp = arr[left]
    arr[left] = arr[right]
    arr[right] = temp

class Solution:
    def partition(self, arr, start, end):
        pivotIndex = randint(start, end)
        swap(arr, pivotIndex, end)
        pivot = arr[end]
        i = start
        for currInd in range(start, end):
            if arr[currInd] <= pivot:
                swap(arr, currInd, i)
                i += 1
        swap(arr, i, end)
        return i

    def improvedPartition(self, arr, start, end):
        # all the elements lesser than the pivot will be in left
        # all the elements greater than the pivot will be in end
        # and all the elements equal to the pivot will be in the middle
        # this algorithm is good for handling duplicate keys
        randomNumber = randint(start, end)
        # swap end index element with  the random index
        swap(arr, end, randomNumber)
        pivot = arr[end]
        fowardIndex = start
        # forward pass to scan elements that are lesser than the pivot
        for i in range(start, end):
            if arr[i] < pivot:
                swap(arr, fowardIndex, i)
                fowardIndex += 1
        # backward pass to scan elements that are greater the pivot
        backwardIndex = end
        for i in range(end, -1, -1):
            if arr[i] > pivot:
                swap(arr, i, backwardIndex)
                backwardIndex -= 1
            # we are already in lesser region
            elif arr[i] < pivot:
                break
        result = [fowardIndex, backwardIndex]
        return result

    def quickSortHelper(self, arr, start, end):
        if start >= end:
            return
        indexes = self.improvedPartition(arr,start, end)
        pStart = indexes[0]
        pEnd = indexes[1]
        self.quickSortHelper(arr, start, pStart - 1)
        self.quickSortHelper(arr, pEnd + 1, end)

    def quickSort(self, arr):
        self.quickSortHelper(arr, 0, len(arr) - 1)
        return arr


arr = [70, 5, 10, 6, 4, 6, 8, 1]
ob = Solution()
print(ob.quickSort(arr))
# for i in range(3, -1, -1):
#     print(i)
ob.improvedPartition([0, 1, 2, 0, 2, 1, 1], 0, 6)
